modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+package proto
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/errors"
+)
+
+// merkleRootsCacheHeight is the height of the subTrees in cachedSubTrees. A
+// height of 7 means that 128 sector roots are covered by a single cached
+// subTree.
+const merkleRootsCacheHeight = 7
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract. The cached trees are not
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	cachedSubTree struct {
+		height int
+		sum    crypto.Hash
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (mr *merkleRoots, err error) {
+	mr = &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(rootIndexToOffset(0), io.SeekStart); err != nil {
+		return
+	}
+	// Read the roots from the file without reading all of them at once.
+	for i := 0; i < mr.numMerkleRoots; i++ {
```

Comment:
I would add a TODO at the top of the file, reading the roots one at a time is pretty inefficient. The OS will handle a lot of that via disk caching, but I'm guessing we would see some i/o and performance improvements if we read say, 100,000 roots at a time instead of one at a time.

No need to do that in this PR, I'd just write a comment highlighting that it's something to do.

Actually, this is the type of optimization that would be really great to throw at a newer developer like Matt or one of the guys in the #core-dev channel.

Reply:
@tbenz9 maybe we should make an issue when this is merged and get someone in the open source community to look at it.

Reply:
optimizing this is really simple thanks to `bufio`:
```go
r := bufio.NewReader(file)
```
You can get a little faster by using a custom (larger) buffer size, but the default buffer size will probably get you to 90% peak performance.